parsers/US_IPC.py

```diff
@@ -0,0 +1,145 @@
+#!usr/bin/env python3
+
+"""Parser for the Idaho Power Comapny area of the United States."""
+
+from dateutil import parser, tz
+from itertools import groupby
+from logging import getLogger
+import requests
+
+
+# NOTE No pumped storage yet but future ideas can be found at the following url.
+# https://docs.idahopower.com/pdfs/AboutUs/PlanningForFuture/irp/IRP.pdf
+
+PRODUCTION_URL = 'https://api.idahopower.com/Energy/Api/V1/GenerationAndDemand/Subset'
+
+GENERATION_MAPPING = {'Non-Utility Geothermal': 'geothermal',
+                      'Hydro': 'hydro',
+                      'Coal': 'coal',
+                      'Diesel': 'oil',
+                      'PURPA/Non-Utility Wind': 'wind',
+                      'Natural Gas': 'gas',
+                      'PURPA/Non-Utility Solar': 'solar',
+                      'PURPA Other': 'biomass'
+                      }
+
+
+def get_data(session=None):
+    """Returns a list of dictionaries."""
+
+    s = session or requests.Session()
+
+    req = requests.get(PRODUCTION_URL)
+    json_data = req.json()
+    raw_data = json_data['list']
+
+    return raw_data
+
+
+def timestamp_converter(timestamp):
+    """Converts a str timestamp into an aware datetime object."""
+
+    dt_naive = parser.parse(timestamp)
+    mountain_time = tz.gettz("America/Boise")
+    dt_aware = dt_naive.replace(tzinfo=dt_naive.tzinfo or mountain_time)
+
+    return dt_aware
+
+
+def data_processer(raw_data, logger):
+    """
+    Groups dictionaries by datetime key.
+    Removes unneeded keys and logs any new ones.
+    Returns a list of tuples containing (datetime object, dictionary).
+    """
+
+    dt_key = lambda x: x['datetime']
+    grouped = groupby(raw_data, dt_key)
+
+    unmapped = set()
+    parsed_data = []
+    for group in grouped:
+        dt = timestamp_converter(group[0])
+        generation = group[1]
+
+        production = {}
+        for gen_type in generation:
+            production[gen_type['name']] = float(gen_type['data'])
+
+        keys_to_remove = {'Load', 'Net Purchases', 'Inadvertent'}
+        known_keys = GENERATION_MAPPING.keys() | set() | keys_to_remove
+        current_keys = production.keys() | set()
+
+        unknown_keys = current_keys - known_keys
+        keys_to_remove = keys_to_remove | unknown_keys
+        unmapped = unmapped | unknown_keys
+
+        for key in keys_to_remove:
+            production.pop(key)
+
+        production = {GENERATION_MAPPING[k]: v for k, v in production.items()}
+
+        parsed_data.append((dt, production))
+
+    for key in unmapped:
+        logger.warning('Key \'{}\' in US-IPC is not mapped to type.'.format(key), extra={'key': 'US-IPC'})
+
+    return parsed_data
+
+
+def fetch_production(zone_key = 'US-SPP', session=None, target_datetime=None, logger=getLogger(__name__)):
```

Comment:
zone_key probably should be `US-IPC` here, or even hardcode it to US-IPC in the function itself. Right now the testing function returns data with `'zoneKey': 'US-SPP'`. This won't be a problem in practice since the backend will call the parser with the right parameter, but it's a bit odd...

Reply:
No idea how I missed that.

Reply:
No worries, that is why we have peer reviews :+1: